ProcessOptimizer/space/space.py

```diff
@@ -1,917 +1,940 @@
+            )
+
+    if len(dimension) == 3:
+        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and dimension[
+            2
+        ] in ["uniform", "log-uniform"]:
+            return Real(*dimension, transform=transform)
+        else:
+            return Categorical(dimension, transform=transform)
+
+    if len(dimension) > 3:
+        return Categorical(dimension, transform=transform)
+
+    raise ValueError(
+        "Invalid dimension {}. Read the documentation for "
+        "supported types.".format(dimension)
+    )
+
+
+class Dimension(ABC):
+    """Base class for search space dimensions."""
+
+    prior = None
+
+    def transform(self, X):
+        """Transform samples form the original space to a warped space."""
+        return self.transformer.transform(X)
+
+    def inverse_transform(self, Xt):
+        """Inverse transform samples from the warped space back into the
+        original space.
+        """
+        return self.transformer.inverse_transform(Xt)
+
+    @property
+    def size(self):
+        return 1
+
+    @property
+    def transformed_size(self):
+        return 1
+
+    @property
+    def bounds(self):
+        raise NotImplementedError
+
+    @property
+    def transformed_bounds(self):
+        raise NotImplementedError
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        if isinstance(value, str) or value is None:
+            self._name = value
+        else:
+            raise ValueError("Dimension's name must be either string or None.")
+
+    def sample(
+        self, points: Union[float, Iterable[float]], allow_duplicates: bool = True
+    ) -> np.ndarray:
+        """Draw points from the dimension.
+
+
+
+        Parameters
+        ----------
+        * `points` [float or list[float]]:
+            A single point or a list of points to sample. All must be between 0 and 1.
+
+        * `allow_duplicates` [bool, default=True]:
+            If True, the output will have the same size as `points`. If False, each
+            point in the output will be unique. This means that the output can be
+            shorter than `points`.
+        """
+        if isinstance(points, (int, float)):  # If a single point is given, convert it
+            # to a list.
+            points = [points]
+        if any([point < 0 or point > 1 for point in points]):
+            raise ValueError("Sample points must be between 0 and 1.")
+        sampled_points = self._sample(points)
+        if not allow_duplicates:
+            # np.unique sorts the inputs, which we do not want, so we have to reinvent
+            # the wheel.
+            seen = set()
+            deduplicated_points = []
+            for point in sampled_points:
+                if point not in seen:
+                    deduplicated_points.append(point)
+                    seen.add(point)
+            sampled_points = deduplicated_points
```

Comment:
```suggestion
            unique_points = []
            for point in sampled_points:
                if point not in seen:
                    unique_points.append(point)
                    seen.add(point)
            sampled_points = unique_points
```
"deduplicated" is rather long way of writing unique